File: packages/fangyuan-rule-parser/test/RuleTest.java
/*
* 此文件由规则文件自动生成请勿编辑
*/
package com.olymtech;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.olymtech.ConcurrentRule;


public class RuleTest extends AbstractRule<List<String>> {

    RuleTest(Rule<List<String>> rule) {
        super(rule);
    }

    static RuleTest getInstance(){
        Map<String, Rule<String>> rules$0 = new HashMap<>();

        rules$0.put("华南地区", new SimpleRule<>(
                "(user.orgName  == \"华南公司\" && user.roleName  == \"航管\") && (inArrayOrObject(feeItem.orgId , [\"1223\",\"123\"]))",
                "org.javatuples.Pair.with(0, \"\")",
                String.class
        ));

        Map<String, Rule<String>> rules$1 = new HashMap<>();
        rules$0.put("非业务主体", new SortRule<>(rules$1));

        rules$1.put("非业务主体锁定", new SimpleRule<>(
                "$orgIds = user.orgIds;(user.orgIds ) && (( !match(feeItem.id , \"^add-\") ) && ( !inArrayOrObject(feeItem.orgId , $userList) ))",
                "$orgIds = user.orgIds;org.javatuples.Pair.with(1, $orgIds)",
                String.class
        ));

        rules$0.put("子单锁定", new SimpleRule<>(
                "(feeItem.splitBillType  == \"child\")",
                "org.javatuples.Pair.with(1, \"子单锁定\")",
                String.class
        ));


        return new RuleTest(new ConcurrentRule<>(rules$0));
    }
}

File: packages/fangyuan-rule-parser/test/RuleTest.ts
/*
 * 此文件由规则文件自动生成请勿编辑
 */

const isDev = process.env.NODE_ENV !== 'production';
// eslint-disable-next-line
const debug = console.debug;
// prettier-ignore
import User from '@/knownledage/User'
import FeeItem from '@/knownledage/FeeItem';
type Context = { user: User; feeItem: FeeItem };

type Rule = (ctx: Context) => {
  evaluate: () => boolean;
  execute: () => [number, string];
};
function inArrayOrObject(key: string, data: string[] | Object) {
  return Array.isArray(data) ? data.includes(key) : key in data;
}
function match(matchString: string, regex: string) {
  return new RegExp(regex).test(matchString);
}
function concurrent(rules: Record<string, Rule>) {
  return (ctx: Context) => {
    let validRules: [string, ReturnType<Rule>][] | undefined;

    function evaluate() {
      validRules = Object.keys(rules)
        .map((key) => [key, rules[key](ctx)] as [string, ReturnType<Rule>])
        .filter(([key, obj]) => {
          const result = obj.evaluate();
          // isDev && debug(key + ":", result)
          return result;
        });
      return validRules.length > 0;
    }

    function execute() {
      if (!validRules) {
        throw new Error('please confirm evaluate is true');
      }
      const result = validRules.map(([, obj]) => obj.execute());
      // isDev && debug("result:", result)
      return result;
    }

    return { evaluate, execute };
  };
}

const rules$0: Record<string, Rule> = (exports.rules$0 = {});
rules$0['华南地区'] = function (ctx: Context) {
  const { user, feeItem } = ctx;
  function evaluate() {
    return (
      user.orgName == '华南公司' &&
      user.roleName == '航管' &&
      inArrayOrObject(feeItem.orgId, ['1223', '123'])
    );
  }
  function execute(): [number, string] {
    return [0, ''];
  }

  return { evaluate, execute };
};
rules$0['非业务主体锁定'] = function (ctx: Context) {
  const { user, feeItem } = ctx;
  let $orgIds = user.orgIds;
  let $user = user;
  function evaluate() {
    return (
      user.orgIds &&
      !match(feeItem.id, '^add-') &&
      !inArrayOrObject(feeItem.orgId, $orgIds)
    );
  }
  function execute(): [number, string] {
    $user.orgIds = 1;
    return [1, '非业务主体锁定'];
  }

  return { evaluate, execute };
};
rules$0['子单锁定'] = function (ctx: Context) {
  const { feeItem } = ctx;
  function evaluate() {
    return feeItem.splitBillType == 'child';
  }
  function execute(): [number, string] {
    return [1, '子单锁定'];
  }

  return { evaluate, execute };
};

module.exports = concurrent(rules$0);
